torch/_lobpcg.py

```diff
@@ -0,0 +1,671 @@
+            M = _utils.qform(_utils.qform(self.A, S_), Ri)
+            E_, Z = _utils.symeig(M, largest)
+            self.X[:, nc:] = mm(S_, mm(Ri, Z[:, nc:n]))
+            self.E[nc:] = E_[:n - nc]
+            P = mm(S_, mm(Ri, Z[:, n:2 * n - nc]))
+            np = P.shape[-1]
+
+            self.update_residual()
+            nc = self.update_converged_count()
+            self.S[..., :n] = self.X
+            self.S[:, n:n + np] = P
+            W = _utils.matmul(self.iK, self.R[:, nc:])
+
+            self.ivars['converged_end'] = ns = n + np + W.shape[-1]
+            self.S[:, n + np:ns] = W
+
+    def _update_ortho(self):
+        """
+        Update or initialize iteration variables when `method == "ortho"`.
+        """
+        mm = torch.matmul
+        ns = self.ivars['converged_end']
+        nc = self.ivars['converged_count']
+        n = self.iparams['n']
+        largest = self.bparams['largest']
+
+        if self.ivars['istep'] == 0:
+            Ri = self._get_rayleigh_ritz_transform(self.X)
+            M = _utils.qform(_utils.qform(self.A, self.X), Ri)
+            E, Z = _utils.symeig(M, largest)
+            self.X = mm(self.X, mm(Ri, Z))
+            self.update_residual()
+            np = 0
+            nc = self.update_converged_count()
+            self.S[:, :n] = self.X
+            W = self._get_ortho(self.R, self.X)
+            ns = self.ivars['converged_end'] = n + np + W.shape[-1]
+            self.S[:, n + np:ns] = W
+
+        else:
+            S_ = self.S[:, nc:ns]
+            # Rayleigh-Ritz procedure
+            E_, Z = _utils.symeig(_utils.qform(self.A, S_), largest)
+
+            # Update E, X, P
+            self.X[:, nc:] = mm(S_, Z[:, :n - nc])
+            self.E[nc:] = E_[:n - nc]
+            P = mm(S_, mm(Z[:, n - nc:], _utils.basis(_utils.transpose(Z[:n - nc, n - nc:]))))
+            np = P.shape[-1]
+
+            # check convergence
+            self.update_residual()
+            nc = self.update_converged_count()
+
+            # update S
+            self.S[:, :n] = self.X
+            self.S[:, n:n + np] = P
+            W = self._get_ortho(self.R[:, nc:], self.S[:, :n + np])
+            ns = self.ivars['converged_end'] = n + np + W.shape[-1]
+            self.S[:, n + np:ns] = W
+
+    def _get_rayleigh_ritz_transform(self, S):
+        """Return a transformation matrix that is used in Rayleigh-Ritz
+        procedure for reducing a general eigenvalue problem :math:`(S^TAS)
+        C = (S^TBS) C E` to a standard eigenvalue problem :math: `(Ri^T
+        S^TAS Ri) Z = Z E` where `C = Ri Z`.
+
+        Arguments:
+        S (Tensor): the matrix basis for the search subspace, size is
+                    :math:`(m, n)`.
+
+        Returns:
+        Ri (tensor): upper-triangular transformation matrix of size
+                     :math:`(n, n)`.
+        """
+        B = self.B
+        mm = torch.matmul
+        SBS = _utils.qform(B, S)
+        d1 = SBS.diagonal(0, -2, -1) ** -0.5
```

Comment:
This step could be optimized, if `d1` is a diagonal matrix, then `d_{ii}` is just an inner product of the `i`-th column of `S` with the `i`-th column of `BS`. Hence this step complexity could be reduced by a factor of `m`.

Reply:
@vincentqb, could you please tell whether there is a difference in computing `BS` versus `S^T * B^T`?

Reply:
Operations are usually faster when accessing contiguous element of a dense tensor,  and so accessing elements through a transpose can be more expensive.

Reply:
I see, I thought that TensorIterator inside kind of does transposition if necessary. Will check then..

Reply:
I see, I thought that TensorIterator inside kind of does transposition if necessary. Will check then..

Reply:
We need full `S^T B S`, so `diag(S^T B S)` is freely available as `d1`.